File: ios/BiblosSurfer/Views/Library/LibraryViewModel.swift
//
//  LibraryViewModel.swift
//  BiblosSurfer
//
//  Created by Mikołaj Linczewski on 21/08/2026.
//

import Foundation
import Observation

protocol LibraryRouting: AnyObject {
    func openBook(_ item: LibraryItem)
}

struct LibraryViewState: Observable {
    var items: [LibraryItem] = []
    var isLoading = true
    var error: DescriptiveError?
}

@Observable
final class LibraryViewModel: ErrorDismissing {
    private let libraryService: LibraryServiceProtocol
    weak var router: LibraryRouting?

    var viewProperties: LibraryViewState

    init(libraryService: LibraryServiceProtocol = LibraryService()) {
        self.libraryService = libraryService
        self.viewProperties = .init()
    }

    @MainActor
    func reload() async {
        viewProperties = .init()
        await load()
    }

    @MainActor
    func load() async {
        viewProperties.isLoading = true
        defer { viewProperties.isLoading = false }
        do {
            viewProperties.items = try await libraryService.loadItems()
            viewProperties.error = nil
        } catch let error as DescriptiveError {
            viewProperties.error = error
        } catch {
            viewProperties.error = Errors.Library.copyFailed(
                fileName: "library",
                underlying: error.localizedDescription
            )
        }
    }

    func open(_ item: LibraryItem) {
        router?.openBook(item)
    }

    @MainActor
    func dismissError() {
        Task { await reload() }
    }

    @MainActor
    func importBook(from url: URL) async {
        do {
            _ = try await libraryService.importBook(from: url)
            await load()
        } catch let error as DescriptiveError {
            viewProperties.error = error
        } catch {
            viewProperties.error = Errors.Library.copyFailed(
                fileName: url.lastPathComponent,
                underlying: error.localizedDescription
            )
        }
    }

    func presentImportFailure(_ error: Error) {
        viewProperties.error = Errors.Library.copyFailed(
            fileName: "import",
            underlying: error.localizedDescription
        )
    }
}

File: android/app/src/main/java/miko/biblossurfer/ui/library/LibraryViewModel.kt
package miko.biblossurfer.ui.library

import miko.biblossurfer.data.DescriptiveError
import miko.biblossurfer.data.Errors
import miko.biblossurfer.data.LibraryServiceProtocol
import miko.biblossurfer.data.model.LibraryItem
import java.io.File
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

interface LibraryRouting {
    fun openBook(item: LibraryItem)
}

interface ErrorDismissing {
    fun dismissError()
}

data class LibraryViewState(
    val items: List<LibraryItem> = emptyList(),
    val isLoading: Boolean = true,
    val error: DescriptiveError? = null,
)

class LibraryViewModel(
    private val libraryService: LibraryServiceProtocol,
) : ErrorDismissing {
    var router: LibraryRouting? = null

    private val _viewProperties = MutableStateFlow(LibraryViewState())
    val viewProperties: StateFlow<LibraryViewState> = _viewProperties.asStateFlow()

    suspend fun reload() {
        _viewProperties.value = LibraryViewState()
        load()
    }

    suspend fun load() {
        _viewProperties.update { it.copy(isLoading = true) }
        try {
            val items = libraryService.loadItems()
            _viewProperties.update { it.copy(items = items, error = null, isLoading = false) }
        } catch (error: Exception) {
            if (error is DescriptiveError) {
                _viewProperties.update { it.copy(error = error, isLoading = false) }
                return
            }
            _viewProperties.update {
                it.copy(
                    error = Errors.Library.CopyFailed("library", error.localizedMessage ?: error.toString()),
                    isLoading = false,
                )
            }
        }
    }

    fun open(item: LibraryItem) {
        router?.openBook(item)
    }

    override fun dismissError() {
        // Caller launches reload on the Main dispatcher, matching iOS `Task { await reload() }`.
    }

    suspend fun importBook(from: File) {
        try {
            libraryService.importBook(from)
            load()
        } catch (error: Exception) {
            if (error is DescriptiveError) {
                _viewProperties.update { it.copy(error = error) }
                return
            }
            _viewProperties.update {
                it.copy(
                    error = Errors.Library.CopyFailed(from.name, error.localizedMessage ?: error.toString())
                )
            }
        }
    }

    fun presentImportFailure(error: Throwable) {
        _viewProperties.update {
            it.copy(
                error = Errors.Library.CopyFailed(
                    "import",
                    error.localizedMessage ?: error.toString(),
                )
            )
        }
    }
}
